test/test_replace.py

```diff
@@ -0,0 +1,225 @@
+        ),
+        (CD(), {}),
+        pytest.param(
+            "str_obj",
+            {},
+            marks=pytest.mark.xfail(reason="Raise TypeError if obj is not dataclass instances"),
+        ),
+        (
+            OuterDataclass(nested=NestedDataclasses(nested_arg="nested_arg_2", arg1=1, arg2=2.0)),
+            {"nested": {"nested_arg": "nested_arg_2", "arg1": 1, "arg2": 2.0}},
+        ),
+    ],
+)
+def test_replace_nested_dict(config: object, changes: dict):
+    config_replaced = replace(type(config)(), changes)
+    logger.info(config_replaced)
+    assert config == config_replaced
+    assert id(config) != id(config_replaced)
+
+
+@pytest.mark.parametrize(
+    ("config", "changes"),
+    [
+        (CD(c_or_d=D(d=1)), {"__subgroups__@c_or_d": "d", "c_or_d.d": 1}),
+        (CD(c_or_d=C(c=True)), {"__subgroups__@c_or_d": "c", "c_or_d.c": True}),
+        (
+            OuterDataclass(
+                some_arg="some_arg_1", nested=NestedDataclasses(nested_arg="nested_arg_1")
+            ),
+            {"some_arg": "some_arg_1", "nested.nested_arg": "nested_arg_1"},
+        ),
+    ],
+)
+def test_replace_flatten_dict(config: object, changes: dict):
+    config_replaced = replace(type(config)(), changes)
+    assert config == config_replaced
+    assert id(config) != id(config_replaced)
+
+
+@dataclass
+class InnerClass:
+    arg1: int = 0
+    arg2: str = "foo"
+
+
+@dataclass(frozen=True)
+class OuterClass:
+    outarg: int = 1
+    nested: InnerClass = field(default_factory=InnerClass)
+
+
+def test_replace_outerdataclass():
+    changes_1 = {"outarg": 2, "nested.arg1": 1, "nested.arg2": "bar"}
+    changes_2 = {"outarg": 2, "nested": {"arg1": 1, "arg2": "bar"}}
+    c = OuterClass()
+    c1 = replace(c, changes_1)
+    c2 = replace(c, changes_2)
+    assert c1 == c2 == OuterClass(nested=InnerClass(arg1=1, arg2="bar"), outarg=2)
+    assert id(c1) != id(c2)
+
+
+def test_replace_nested_1():
+    @dataclass
+    class C:
+        c: bool = False
+
+    @dataclass
+    class D:
+        d: int = 0
+
+    @dataclass(frozen=True)
+    class CD(TestSetup):
+        c_or_d: C | D = subgroups({"c": C, "d": D}, default="c")
+
+        other_arg: str = "bob"
+
+    config = CD(c_or_d=C())
+    assert replace(config, {"c_or_d": {"c": True}}).c_or_d.c == True
+
+    config = CD(c_or_d=D())
+    assert replace(config, {"c_or_d": {"d": 2}}).c_or_d.d == 2
+
+    config = CD(c_or_d=D())
+    assert replace(config, {"c_or_d": D(d=2)}).c_or_d.d == 2
+
+    config = CD(c_or_d=D())
+    assert replace(config, {"c_or_d": "d"}).c_or_d.d == 0
+
+
+@dataclass
+class Config(TestSetup):
+    ab_or_cd: AB | CD = subgroups(
+        {"ab": AB, "cd": CD},
+        default_factory=AB,
+    )
+
+
+@pytest.mark.parametrize(
+    ("config", "changes"),
+    [
+        (
+            Config(ab_or_cd=CD(c_or_d=D())),
+            {"__subgroups__@ab_or_cd": "cd", "ab_or_cd.__subgroups__@c_or_d": "d"},
```

Comment:
I have an idea, tell me what you think:
At this point, couldn't the `replace` function take an additional special argument called `subgroups: dict[str, str]` that maps from the field names (flattened) to the subgroup to change?
For instance,  `replace(Config(), subgroups={"ab_or_cd": "cd", "ab_or_cd.c_or_d": "d"})`?
Perhaps it could have a special name like `_subgroups`, if it's possible for the dataclass to have a field called `subgroups`.

We could also not add an argument, but instead use the type of `new_values_dict` to decide if we interpret those values as dataclass fields, or as subgroup choices:
```python
class SubgroupDict(dict):
    ...
replace(Config(), SubgroupDict({"ab_or_cd": "cd", "ab_or_cd.c_or_d": "d"}))
```

Even more interesting, we could use a small wrapper to indicate that a value is the subgroup name, rather than a string value:
```python
class SubgroupName:
    def __init__(self, name: Any):
        self.name = name
replace(Config(), {"ab_or_cd": SubgroupName("cd"), "ab_or_cd.c_or_d": SubgroupName("d"), "ab_or_cd.c_or_d.d": 123})
```

Anyways, it's getting late here, and I'm just rambling at this point. Let me know what you think.

Reply:
Those are really good ideas. If I have to pick one from them, I would prefer use the less intrusive way like the small wrapper. But this needs to reconsider for the key conflict problem as I mentioned below.
We could wrapper the `__subgroups__@[key]` in a class that can regularize user's input.